bundles/org.eclipse.swt/Eclipse SWT Custom Widgets/common/org/eclipse/swt/custom/BusyIndicator.java

```diff
@@ -70,6 +70,41 @@ public static void showWhile(Display display, Runnable runnable) {
 		}
 	}
 
+	/**
+	 * If called from a {@link Display} thread, waits for the given
+	 * <code>Future</code> to complete and provides busy feedback using the
+	 * busy indicator. While waiting for completion, pending UI events are processed
+	 * to prevent UI freeze.
+	 *
+	 * If there is no {@link Display} for the current thread, the method returns
+	 * immediately and no feedback is shown to the user.
+	 *
+	 * In any case, the passed {@link Future} is returned for chaining of further operations.
+	 *
+	 * @param future the {@link Future} for which busy feedback is to be shown.
+	 * @return the {@link Future} for chaining.
+	 * @since 3.127
+	 */
+	public static <F extends Future<T>, T> F showWhile(F future) {
+		Objects.requireNonNull(future);
+		if (!future.isDone()) {
+			Display display = Display.findDisplay(Thread.currentThread());
+			if (display != null) {
+				Integer busyId = setBusyCursor(display);
+				try {
+					while (!future.isDone() && !display.isDisposed()) {
+						if (!display.readAndDispatch()) {
+							Thread.onSpinWait();
+						}
+					}
+				} finally {
+					clearBusyCursor(display, busyId);
```

Comment:
What if this method is called multiple times with different futures?

As I currently understand it, this could lead to unexpected result:
1. firstFuture is created and starts
2. showWile(firtFuture)  => busy cursor
3. secondFuture is created and starts
4. showWhile(secondFuture)
5. firstFuture completes => clearBusyCursor(display, busyId) => normal cursor
6. secondFuture is still running

Note that this is an issue with BusyIndicator in general. But it might be easier to fix it with futures.

Reply:
> Note that this is an issue with BusyIndicator in general. But it might be easier to fix it with futures.

I just want to enhance it so it can be used easier and safe as the "classic" implementation is hard to use and prone to UI freeze because one has to guess how it is to be used (that is: how I implemented it now with the futures), this is nothing to really blame on the original authors because at that time CompletableFuture, maybe even the nice Java Concurrent things we have today where not part of Java at all.

I just don't want to fix general problems, if there are any, please create an issue with a testcase that shows the problem and it can probably be solved (independent from that).

In general SWT allows many "worse" combinations of syn/asnyc/nested/whatever calls, we can't solve them all we can just give better building blocks.

Reply:
I now aligned it with the "traditional" `BusyIndicator.showWhile(Display, Runnable)` by throwing SWT error...